File: yntra-ios/YntraIOS/Views/TimeClockView.swift
import SwiftUI
import yntra_core

struct TimeClockView: View {
    @StateObject private var locationManager = LocationManager()
    @State private var userId: String = "usr_field_01"
    @State private var workspaceId: String = "ws_default"
    @State private var hours: String = "8.0"
    @State private var shiftNote: String = "Field job shift"
    @State private var clockStatusMessage: String? = nil
    @State private var isClockedIn: Bool = false
    @State private var activeShiftStartTime: Date? = nil
    
    private let darkBackground = Color(red: 0.04, green: 0.06, blue: 0.1)
    private let cardBackground = Color(red: 0.12, green: 0.16, blue: 0.23)
    private let primaryPurple = Color(red: 0.55, green: 0.36, blue: 0.96)
    
    var body: some View {
        ZStack {
            darkBackground.ignoresSafeArea()
            
            VStack(alignment: .leading, spacing: 20) {
                // Header
                HStack {
                    VStack(alignment: .leading, spacing: 4) {
                        Text("GPS Time Clock")
                            .font(.system(size: 28, weight: .black, design: .rounded))
                            .foregroundColor(.white)
                        Text("Geofenced Field Shift Clocking")
                            .font(.system(size: 13, weight: .medium, design: .rounded))
                            .foregroundColor(.white.opacity(0.6))
                    }
                    Spacer()
                }
                .padding(.horizontal)
                .padding(.top, 10)
                
                // Geofence Badge & Location Card
                VStack(alignment: .leading, spacing: 12) {
                    HStack {
                        Image(systemName: locationManager.isWithinGeofence ? "checkmark.seal.fill" : "location.slash.fill")
                            .foregroundColor(locationManager.isWithinGeofence ? .green : .orange)
                            .font(.system(size: 20))
                        
                        VStack(alignment: .leading, spacing: 2) {
                            Text(locationManager.isWithinGeofence ? "Inside Job Site Geofence" : "Outside Geofence Radius")
                                .font(.system(size: 15, weight: .bold, design: .rounded))
                                .foregroundColor(.white)
                            
                            Text("Distance: \(String(format: "%.1f", locationManager.distanceFromTargetMeters))m from site (Limit: \(Int(locationManager.maxRadiusMeters))m)")
                                .font(.system(size: 12, weight: .medium, design: .rounded))
                                .foregroundColor(.white.opacity(0.6))
                        }
                        Spacer()
                    }
                    
                    if let userLoc = locationManager.userLocation {
                        Text("Coordinates: \(userLoc.coordinate.latitude), \(userLoc.coordinate.longitude)")
                            .font(.system(size: 11, weight: .monospaced))
                            .foregroundColor(.white.opacity(0.4))
                    }
                }
                .padding(16)
                .background(cardBackground)
                .cornerRadius(16)
                .overlay(
                    RoundedRectangle(cornerRadius: 16)
                        .stroke(locationManager.isWithinGeofence ? Color.green.opacity(0.3) : Color.orange.opacity(0.3), lineWidth: 1)
                )
                .padding(.horizontal)
                
                if let statusMsg = clockStatusMessage {
                    Text(statusMsg)
                        .font(.system(size: 13, weight: .medium, design: .rounded))
                        .foregroundColor(statusMsg.contains("Violation") || statusMsg.contains("Failed") ? .red : .green)
                        .padding(.horizontal)
                }
                
                // Clock Controls Card
                VStack(alignment: .leading, spacing: 16) {
                    Text("SHIFT DETAILS")
                        .font(.system(size: 11, weight: .bold))
                        .foregroundColor(.white.opacity(0.5))
                    
                    TextField("Shift Notes...", text: $shiftNote)
                        .textFieldStyle(PlainTextFieldStyle())
                        .padding(12)
                        .background(darkBackground)
                        .cornerRadius(10)
                        .foregroundColor(.white)
                    
                    TextField("Logged Hours (e.g. 8.0)", text: $hours)
                        .textFieldStyle(PlainTextFieldStyle())
                        .padding(12)
                        .background(darkBackground)
                        .cornerRadius(10)
                        .foregroundColor(.white)
                    
                    Button(action: {
                        triggerClockIn()
                    }) {
                        HStack {
                            Image(systemName: isClockedIn ? "stop.circle.fill" : "play.circle.fill")
                                .font(.system(size: 22))
                            Text(isClockedIn ? "Clock Out & Submit Shift" : "GPS Clock In")
                                .font(.system(size: 16, weight: .bold, design: .rounded))
                        }
                        .foregroundColor(.white)
                        .frame(maxWidth: .infinity)
                        .padding(.vertical, 16)
                        .background(isClockedIn ? Color.red : (locationManager.isWithinGeofence ? primaryPurple : Color.gray))
                        .cornerRadius(14)
                    }
                    .disabled(!locationManager.isWithinGeofence && !isClockedIn)
                }
                .padding(18)
                .background(cardBackground)
                .cornerRadius(18)
                .padding(.horizontal)
                
                Spacer()
            }
        }
        .onAppear {
            locationManager.requestLocationPermission()
        }
    }
    
    private func triggerClockIn() {
        guard let userLoc = locationManager.userLocation else {
            clockStatusMessage = "Failed: Device GPS location unavailable"
            return
        }
        
        let hrs = Double(hours) ?? 8.0
        let dateStr = DateFormatter.yyyyMMdd.string(from: Date())
        
        Task {
            do {
                if !isClockedIn {
                    let report = try await clockInGeofenced(
                        requesterUserId: userId,
                        workspaceId: workspaceId,
                        userId: userId,
                        teamId: nil,
                        date: dateStr,
                        hours: hrs,
                        note: shiftNote,
                        latitude: userLoc.coordinate.latitude,
                        longitude: userLoc.coordinate.longitude,
                        targetLatitude: locationManager.targetLatitude,
                        targetLongitude: locationManager.targetLongitude,
                        maxRadiusMeters: locationManager.maxRadiusMeters,
                        isSpoofedLocation: locationManager.isLocationSpoofed,
                        polygonCoordsJson: nil
                    )
                    isClockedIn = true
                    activeShiftStartTime = Date()
                    clockStatusMessage = "Clock-in verified! Report ID: \(report.id)"
                } else {
                    isClockedIn = false
                    clockStatusMessage = "Clock-out completed successfully."
                }
            } catch {
                clockStatusMessage = error.localizedDescription
            }
        }
    }
}

extension DateFormatter {
    static let yyyyMMdd: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter
    }()
}

File: yntra-android/app/src/main/java/com/yntra/app/views/TimeClockView.kt
package com.yntra.app.views

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.LocationOff
import androidx.compose.material.icons.filled.PlayArrow
import androidx.compose.material.icons.filled.Stop
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.yntra.app.utils.AndroidLocationManager
import kotlinx.coroutines.launch
import uniffi.yntra_core.clockInGeofenced

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun TimeClockView() {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    val locationManager = remember { AndroidLocationManager(context) }
    val locationState by locationManager.locationState.collectAsState()

    var hoursText by remember { mutableStateOf("8.0") }
    var noteText by remember { mutableStateOf("Field Shift Work") }
    var isClockedIn by remember { mutableStateOf(false) }
    var statusMessage by remember { mutableStateOf<String?>(null) }

    DisposableEffect(Unit) {
        locationManager.startLocationUpdates()
        onDispose {
            locationManager.stopLocationUpdates()
        }
    }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(Color(0xFF0B0F19))
            .padding(20.dp),
        verticalArrangement = Arrangement.spacedBy(16.dp)
    ) {
        Text(
            text = "GPS Time Clock",
            color = Color.White,
            fontSize = 26.sp,
            fontWeight = FontWeight.Black
        )
        Text(
            text = "Geofenced Field Shift Clocking",
            color = Color(0xFF94A3B8),
            fontSize = 13.sp
        )

        // Geofence Badge & Location Card
        Card(
            colors = CardDefaults.cardColors(containerColor = Color(0xFF1E293B)),
            shape = RoundedCornerShape(16.dp),
            modifier = Modifier
                .fillMaxWidth()
                .border(
                    width = 1.dp,
                    color = if (locationState.isWithinGeofence) Color(0xFF10B981) else Color(0xFFF59E0B),
                    shape = RoundedCornerShape(16.dp)
                )
        ) {
            Column(
                modifier = Modifier.padding(16.dp),
                verticalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(12.dp)
                ) {
                    Icon(
                        imageVector = if (locationState.isWithinGeofence) Icons.Default.CheckCircle else Icons.Default.LocationOff,
                        contentDescription = "Geofence Status",
                        tint = if (locationState.isWithinGeofence) Color(0xFF10B981) else Color(0xFFF59E0B)
                    )
                    Column {
                        Text(
                            text = if (locationState.isWithinGeofence) "Inside Job Site Geofence" else "Outside Geofence Radius",
                            color = Color.White,
                            fontWeight = FontWeight.Bold,
                            fontSize = 15.sp
                        )
                        Text(
                            text = "Distance: %.1fm from target site".format(locationState.distanceFromTargetMeters),
                            color = Color(0xFF94A3B8),
                            fontSize = 12.sp
                        )
                    }
                }

                if (locationState.latitude != 0.0) {
                    Text(
                        text = "GPS: %.5f, %.5f".format(locationState.latitude, locationState.longitude),
                        color = Color(0xFF64748B),
                        fontSize = 11.sp
                    )
                }
            }
        }

        statusMessage?.let { msg ->
            Text(
                text = msg,
                color = if (msg.contains("Violation") || msg.contains("Failed")) Color.Red else Color.Green,
                fontSize = 13.sp,
                fontWeight = FontWeight.Medium
            )
        }

        // Shift Details Card
        Card(
            colors = CardDefaults.cardColors(containerColor = Color(0xFF1E293B)),
            shape = RoundedCornerShape(18.dp),
            modifier = Modifier.fillMaxWidth()
        ) {
            Column(
                modifier = Modifier.padding(18.dp),
                verticalArrangement = Arrangement.spacedBy(14.dp)
            ) {
                Text(
                    text = "SHIFT DETAILS",
                    color = Color(0xFF64748B),
                    fontSize = 11.sp,
                    fontWeight = FontWeight.Bold
                )

                OutlinedTextField(
                    value = noteText,
                    onValueChange = { noteText = it },
                    label = { Text("Shift Note") },
                    colors = OutlinedTextFieldDefaults.colors(
                        focusedTextColor = Color.White,
                        unfocusedTextColor = Color.White,
                        focusedBorderColor = Color(0xFF8B5CF6),
                        unfocusedBorderColor = Color(0xFF334155)
                    ),
                    modifier = Modifier.fillMaxWidth()
                )

                OutlinedTextField(
                    value = hoursText,
                    onValueChange = { hoursText = it },
                    label = { Text("Logged Hours") },
                    colors = OutlinedTextFieldDefaults.colors(
                        focusedTextColor = Color.White,
                        unfocusedTextColor = Color.White,
                        focusedBorderColor = Color(0xFF8B5CF6),
                        unfocusedBorderColor = Color(0xFF334155)
                    ),
                    modifier = Modifier.fillMaxWidth()
                )

                Button(
                    onClick = {
                        val hrs = hoursText.toDoubleOrNull() ?: 8.0
                        scope.launch {
                            try {
                                if (!isClockedIn) {
                                    val report = clockInGeofenced(
                                        requesterUserId = "usr_field_01",
                                        workspaceId = "ws_default",
                                        userId = "usr_field_01",
                                        teamId = null,
                                        date = "2026-08-04",
                                        hours = hrs,
                                        note = noteText,
                                        latitude = locationState.latitude,
                                        longitude = locationState.longitude,
                                        targetLatitude = 59.3293,
                                        targetLongitude = 18.0686,
                                        maxRadiusMeters = 250.0,
                                        isSpoofedLocation = locationState.isLocationSpoofed,
                                        polygonCoordsJson = null
                                    )
                                    isClockedIn = true
                                    statusMessage = "Clock-in verified! Report ID: ${report.id}"
                                } else {
                                    isClockedIn = false
                                    statusMessage = "Clock-out completed successfully."
                                }
                            } catch (e: Exception) {
                                statusMessage = e.localizedMessage
                            }
                        }
                    },
                    colors = ButtonDefaults.buttonColors(
                        containerColor = if (isClockedIn) Color.Red else (if (locationState.isWithinGeofence) Color(0xFF8B5CF6) else Color.Gray)
                    ),
                    shape = RoundedCornerShape(14.dp),
                    modifier = Modifier.fillMaxWidth(),
                    enabled = locationState.isWithinGeofence || isClockedIn
                ) {
                    Row(
                        horizontalArrangement = Arrangement.spacedBy(8.dp),
                        verticalAlignment = Alignment.CenterVertically,
                        modifier = Modifier.padding(vertical = 4.dp)
                    ) {
                        Icon(
                            imageVector = if (isClockedIn) Icons.Default.Stop else Icons.Default.PlayArrow,
                            contentDescription = null
                        )
                        Text(
                            text = if (isClockedIn) "Clock Out & Submit Shift" else "GPS Clock In",
                            fontSize = 16.sp,
                            fontWeight = FontWeight.Bold
                        )
                    }
                }
            }
        }
    }
}
